File: core/src/main/java/dev/jason/project/spring/vc_server/core/dto/UserDto.java
package dev.jason.project.spring.vc_server.core.dto;

import dev.jason.project.spring.vc_server.core.model.Device;
import dev.jason.project.spring.vc_server.core.model.User;

public record UserDto(String uid, String displayName, String profilePictureUrl, User.Status status, DeviceDto device) {
	
	public UserDto(String uid, String displayName, String profilePictureUrl, User.Status status) {
		this(uid, displayName, profilePictureUrl, status, null);
	}
	
	public static UserDto fromUser(User user) {
		return new UserDto(user.uid(), user.displayName(), user.profilePictureUrl(), user.status());
	}
	
	public static UserDto fromUserAndDevice(User user, Device device) {
		return new UserDto(user.uid(), user.displayName(), user.profilePictureUrl(), user.status(), DeviceDto.asDto(device));
	}
	
	public User asUser() {
		return new User(
            uid,
            displayName,
            profilePictureUrl,
            User.Status.Online,
            System.currentTimeMillis()
        );
	}
}

File: core/src/main/java/dev/jason/app/compose/vaultchat/core/model/UserDto.kt
package dev.jason.app.compose.vaultchat.core.model

import kotlinx.serialization.Serializable

@Serializable
data class UserDto(
    val uid: String,
    val displayName: String,
    val profilePictureUrl: String,
    val status: Status,
    val device: DeviceDto? = null
) {
    @Serializable
    enum class Status {
        Offline, Online
    }
}

fun UserDto.Status.toStatus() = when (this) {
    UserDto.Status.Offline -> User.Status.Offline
    UserDto.Status.Online -> User.Status.Online
}

fun User.Status.toStatus() = when (this) {
    User.Status.Online -> UserDto.Status.Online
    User.Status.Offline -> UserDto.Status.Offline
}

fun UserDto.toUser() = User(
    uid = uid,
    displayName = displayName,
    profilePictureUrl = profilePictureUrl,
    status = status.toStatus()
)

fun User.toDto(device: DeviceDto? = null) = UserDto(
    uid = uid,
    displayName = displayName,
    profilePictureUrl = profilePictureUrl,
    status = status.toStatus(),
    device = device
)